### PR TITLE
compressFirmwareXz: never compress lib/firmware/edid

### PR DESCRIPTION
**Pull request by nazarewk**
this allows us not to pass `compressFirmware = false;` inside EDID derivations as this mechanic very tricky to discover.

related to #279789

## Description of changes

Always skips `edid` directory when doing compression, leaving it intact.

## Things done

I have tested it end-to-end through https://github.com/nazarewk-iac/nix-configs/commit/b9d935fca851e63a719e9bf6242c91bb10e1fe7d by using `nixos-rebuild switch` and inspecting image with `lsinitrd`:

```shell
$ lsinitrd $(cat /boot/loader/entries/*.conf | grep initrd | tail -n 1 | sed 's#initrd /efi#/boot/EFI#g') | grep 'firmware'
drwxr-xr-x   0 root     root            0 Jan  1  1970 ./nix/store/1419nq86gk6fkicq1y847d0ivgikap74-linux-6.6.10-modules-shrunk/lib/firmware/
drwxr-xr-x   0 root     root            0 Jan  1  1970 ./nix/store/1419nq86gk6fkicq1y847d0ivgikap74-linux-6.6.10-modules-shrunk/lib/firmware/edid/
-r--r--r--   0 root     root          128 Jan  1  1970 ./nix/store/1419nq86gk6fkicq1y847d0ivgikap74-linux-6.6.10-modules-shrunk/lib/firmware/edid/PG278Q_120.bin
-r--r--r--   0 root     root          128 Jan  1  1970 ./nix/store/1419nq86gk6fkicq1y847d0ivgikap74-linux-6.6.10-modules-shrunk/lib/firmware/edid/PG278Q_60.bin
-r--r--r--   0 root     root          128 Jan  1  1970 ./nix/store/1419nq86gk6fkicq1y847d0ivgikap74-linux-6.6.10-modules-shrunk/lib/firmware/edid/U2711_60.bin
drwxr-xr-x   0 root     root            0 Jan  1  1970 ./nix/store/1419nq86gk6fkicq1y847d0ivgikap74-linux-6.6.10-modules-shrunk/lib/firmware/rtl_nic/
-r--r--r--   0 root     root          864 Jan  1  1970 ./nix/store/1419nq86gk6fkicq1y847d0ivgikap74-linux-6.6.10-modules-shrunk/lib/firmware/rtl_nic/rtl8105e-1.fw.xz
...
```

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by soupglasses** [comment]
This change should be fine, as EDID files are a static 256 bytes in size. So skipping compressing edid files should not increase initrd size.

**Comment by cole-h** [comment]
@ofborg eval

File: pkgs/build-support/kernel/compress-firmware-xz.nix [hunk 1]
```diff
@@ -7,14 +7,18 @@ let
 in
 
 runCommand "${firmware.name}-xz" args ''
+  edids=(-path 'lib/firmware/edid/*')
   mkdir -p $out/lib
   (cd ${firmware} && find lib/firmware -type d -print0) |
-      (cd $out && xargs -0 mkdir -v --)
-  (cd ${firmware} && find lib/firmware -type f -print0) |
+      (cd $out && xargs -0 mkdir -pv --)
+  (cd ${firmware} && find lib/firmware -type f -not "''${edids[@]}" -print0) |
```
Comment (flokli) [inline pkgs/build-support/kernel/compress-firmware-xz.nix:14]: Why does this need to be a bash array thing?

Comment (nazarewk) [reply pkgs/build-support/kernel/compress-firmware-xz.nix:14]: It's just my preference to use array whenever there is more than 1 argument and running inside bash.

There is no chance to misuse array as opposed to expanding multi-argument variable.

Comment (flokli) [reply pkgs/build-support/kernel/compress-firmware-xz.nix:14]: But you leave the globbing up to `find`, no? So there's no spaces in paths.

I think it'd make more sense for readability to just "constant-propagate" the -path 'lib/firmware/edid/*' into the three occurences, and add a comment in front, stating edids are excluded from compression.

Comment (nazarewk) [reply pkgs/build-support/kernel/compress-firmware-xz.nix:14]: > But you leave the globbing up to `find`, no? So there's no spaces in paths.
This is not really globbing, just a prefix, `find` was not happy about `-path lib/firmware/edid`

Comment (flokli) [reply pkgs/build-support/kernel/compress-firmware-xz.nix]: But it'd be happy about ' -path 'lib/firmware/edid/*', no?

> I think it'd make more sense for readability to just "constant-propagate" the -path 'lib/firmware/edid/*' into the three occurences, and add a comment in front, stating edids are excluded from compression.

I think that's approach is simpler than these bash arrays. We can revisit once there's more than a single element, but let's keep this simple for now, happy to merge then.

Comment (nazarewk) [reply pkgs/build-support/kernel/compress-firmware-xz.nix]: well, let it be, I left 2 notes on top of the file.

**Review by flokli — APPROVED** [review]

**Pull request closed without merging**